File: node-api/src/services/book.service.ts
import { Injectable } from '@nestjs/common'
import { BookRepository } from '../repositories/book.repository'
import { CreateBookDto } from '../models/create-book.dto'
import { UpdateBookDto } from '../models/update-book.dto'

@Injectable()
export class BookService {
  constructor(private readonly bookRepository: BookRepository) {}

  async create(createBookDto: CreateBookDto) {
    try {
      return await this.bookRepository.createBook(createBookDto)
    } catch (error) {
      throw new Error('Erro ao criar livro!')
    }
  }

  async findAll() {
    try {
      return await this.bookRepository.findAllBooks()
    } catch (error) {
      throw new Error('Erro ao buscar livros!')
    }
  }

  async findOne(id: string) {
    try {
      const book = await this.bookRepository.findBookById(id)
      if (!book) {
        throw new Error('Livro não encontrado')
      }
      return book
    } catch (error) {
      throw new Error('Erro ao buscar livro pelo ID!')
    }
  }

  async update(id: string, updateBookDto: UpdateBookDto) {
    try {
      return await this.bookRepository.updateBook(id, updateBookDto)
    } catch (error) {
      throw new Error('Erro ao atualizar livro!')
    }
  }

  async remove(id: string) {
    try {
      return await this.bookRepository.deleteBook(id)
    } catch (error) {
      throw new Error('Erro ao remover livro!')
    }
  }
}

File: go-echo-api/internal/services/book.service.go
package services

import (
	"go-books-api/internal/models"
	"go-books-api/internal/repositories"
)

type BookService struct {
	Repo *repositories.BookRepository
}

func (s *BookService) CreateBook(book models.Book) error {
	return s.Repo.Create(book)
}

func (s *BookService) GetBooks() ([]models.Book, error) {
	return s.Repo.FindAll()
}

func (s *BookService) GetBookByID(id uint) (models.Book, error) {
	return s.Repo.FindByID(id)
}

func (s *BookService) UpdateBook(book models.Book) error {
	return s.Repo.Update(book)
}

func (s *BookService) DeleteBook(id uint) error {
	return s.Repo.Delete(id)
}
